// nx/src/hasher/impl/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hashArray = exports.fileHasher = exports.getHashingImplementation = exports.HasherImplementation = void 0;
const node_file_hasher_1 = require("./node-file-hasher");
const native_file_hasher_1 = require("./native-file-hasher");
var HasherImplementation;
(function (HasherImplementation) {
    HasherImplementation["Native"] = "Native";
    HasherImplementation["Node"] = "Node";
})(HasherImplementation = exports.HasherImplementation || (exports.HasherImplementation = {}));
function getHashingImplementation() {
    try {
        if ((!process.env.NX_NON_NATIVE_HASHER ||
            process.env.NX_NON_NATIVE_HASHER != 'true') &&
            native_file_hasher_1.NativeFileHasher.available()) {
            return HasherImplementation.Native;
        }
        return HasherImplementation.Node;
    }
    catch (_a) {
        return HasherImplementation.Node;
    }
}
exports.getHashingImplementation = getHashingImplementation;
function createFileHasher() {
    switch (getHashingImplementation()) {
        case HasherImplementation.Native:
            return new native_file_hasher_1.NativeFileHasher();
        case HasherImplementation.Node:
            return new node_file_hasher_1.NodeFileHasher();
    }
}
function createHashArray() {
    switch (getHashingImplementation()) {
        case HasherImplementation.Native:
            return native_file_hasher_1.nativeHashArray;
        case HasherImplementation.Node:
            return node_file_hasher_1.nodeHashArray;
    }
}
exports.fileHasher = createFileHasher();
exports.hashArray = createHashArray();
